apns2/client.py

```diff
@@ -1,39 +1,149 @@
+import collections
+import json
+import logging
 from enum import Enum
-from json import dumps
 
-import json
 from hyper import HTTP20Connection
 from hyper.tls import init_context
 
-from apns2.errors import exception_class_for_reason
-
+from .errors import exception_class_for_reason
 
 class NotificationPriority(Enum):
-    Immediate = 10
-    Delayed = 5
+    Immediate = '10'
+    Delayed = '5'
 
+DEFAULT_APNS_PRIORITY = NotificationPriority.Immediate
+CONCURRENT_STREAMS_SAFETY_MAXIMUM = 1000
+
+logger = logging.getLogger(__name__)
 
 class APNsClient(object):
     def __init__(self, cert_file, use_sandbox=False, use_alternative_port=False):
         server = 'api.development.push.apple.com' if use_sandbox else 'api.push.apple.com'
         port = 2197 if use_alternative_port else 443
         ssl_context = init_context()
         ssl_context.load_cert_chain(cert_file)
-        self.__connection = HTTP20Connection(server, port, ssl_context=ssl_context)
+        self._connection = HTTP20Connection(server, port, ssl_context=ssl_context)
+        self._max_concurrent_streams = None
+        self._previous_server_max_concurrent_streams = None
+
+    def send_notification(
+        self,
+        token_hex,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        stream_id = self.send_notification_async(token_hex, notification, topic, priority)
+        result = self.get_notification_result(stream_id)
+        if result != 'Success':
+            raise exception_class_for_reason(result)
 
-    def send_notification(self, token_hex, notification, priority=NotificationPriority.Immediate, topic=None):
-        json_payload = dumps(notification.dict(), ensure_ascii=False, separators=(',', ':')).encode('utf-8')
+    def send_notification_async(
+        self,
+        token_hex,
+        notification,
+        topic,
+        priority=NotificationPriority.Immediate
+    ):
+        json_payload = json.dumps(
+            notification.dict(),
+            ensure_ascii=False,
+            separators=(',', ':')
+        ).encode('utf-8')
 
-        headers = {
-            'apns-priority': priority.value
-        }
-        if topic:
-            headers['apns-topic'] = topic
+        headers = {'apns-topic': topic}
```

Comment:
Also, there's no reason to send priority if it's set to the default value. It just wastes time and bandwidth.